// @airslate/platform-dev-tools/babel.config.main-module.js
// Need improve
// https://github.com/facebook/react/blob/master/babel.config.js
// https://github.com/zloirock/core-js/blob/master/docs/2019-03-19-core-js-3-babel-and-a-look-into-the-future.md
// https://github.com/babel/babel/blob/master/packages/babel-preset-env/package.json
// https://babeljs.io/docs/en/babel-preset-env#targets
// https://babeljs.io/docs/en/babel-plugin-proposal-class-properties
// https://itnext.io/guide-react-app-monorepo-with-lerna-d932afb2e875
// https://github.com/philipwalton/webpack-esnext-boilerplate/blob/master/tasks/bundles.js
// https://philipwalton.com/articles/deploying-es2015-code-in-production-today/
// https://www.freecodecamp.org/news/you-might-not-need-to-transpile-your-javascript-4d5e0a438ca/
// https://medium.com/hackernoon/polyfills-everything-you-ever-wanted-to-know-or-maybe-a-bit-less-7c8de164e423

const NODE_MODULES = 'node_modules';
const pluginsToTranspile = [
  'query-string',
  'split-on-first',
  'strict-uri-encode',
  'superagent',
  'array-move',
  'serialize-error',
];
const SRC = 'src';

const checker = (dirname) => dirname.includes(NODE_MODULES)
  && pluginsToTranspile.every((p) => !dirname.includes(p))
  && !dirname.includes(SRC);

module.exports = ({
  exclude: [
    checker,
    /[/\\]locales[/\\]en/,
  ],
  sourceType: 'unambiguous',
  presets: [
    ['@babel/env', {
      targets: {
        browsers: [
          'last 2 Chrome versions', 'not Chrome < 60',
          'last 2 Safari versions', 'not Safari < 10.1',
          'last 2 iOS versions', 'not iOS < 10.3',
          'last 2 Firefox versions', 'not Firefox < 54',
          'last 2 Edge versions', 'not Edge < 15',
          'ie 11',
        ],
      },
      modules: false,
      useBuiltIns: 'usage',
      corejs: { version: 3, proposals: true },
      loose: false,
    }],
    '@babel/typescript',
    '@babel/react',
  ],
  plugins: [
    ['@babel/transform-runtime', { regenerator: true }],
    '@babel/proposal-class-properties',
  ],
  env: {
    production: {
      plugins: [
        ['transform-react-remove-prop-types', { removeImport: true }],
      ],
    },
  },
});
