// @airslate/platform-eslint/eslint/rules/no-src-import.js
/* eslint-disable @typescript-eslint/no-var-requires */
const mm = require('micromatch');

module.exports = {
  meta: {
    type: 'suggestion',
    docs: {
      description: 'Prevents importing files from the src directory',
      category: 'Possible Errors',
      suggestion: 'Please use allowed import directories like (@airslate/platform-test/Button)',
    },
    schema: {
      type: 'array',
      items: {
        type: 'string',
      },
    },
  },
  create({ options, report }) {
    const notify = (node) => {
      report({
        node,
        message: `Import should not be done from \'src\' directory(${node.source.value})`,
        data: {
          identifier: node.name,
        },
      });
    };
    return {
      ImportDeclaration(node) {
        const { value } = node.source;

        const srcImport = mm.contains(value, '/src') && !mm.contains(value, '../src');
        const isOptionsMatch = mm.contains(value, options);

        if (options.length) {
          if (srcImport && isOptionsMatch) {
            notify(node);
          }
        } else if (srcImport) {
          notify(node);
        }
      },
    };
  },
};
