// @airslate/platform-eslint/.eslintrc.js
module.exports = {
  env: {
    browser: true,
    es6: true,
    node: true,
  },
  extends: [
    'airbnb',
    'plugin:jest/recommended',
    'plugin:@typescript-eslint/recommended',
  ],
  globals: {
    PLATFORM: 'readonly',
    __LOCAL__: false,
    __DEV__: false,
    __PROD__: false,
    __APP_NAME__: false,
    __BRANCH__: false,
    __BUILD_NUMBER__: false,
    __SENTRY_KEY__: false,
    __TARGET__: false,
    __AUTO_AUTH__: false,
    __RELEASE__: 'readonly',
    __PLATFORM_SDK_CONFIG__: 'readonly',
  },
  parser: '@typescript-eslint/parser',
  parserOptions: {
    ecmaVersion: 7,
    sourceType: 'module',
    ecmaFeatures: {
      jsx: true,
    },
  },
  overrides: [
    {
      files: ['**/*.tsx'],
      rules: {
        'react/prop-types': 'off',
      },
    },
  ],
  plugins: [
    'import',
    'jsx-a11y',
    'react',
    'react-hooks',
    // 'platform-rules',
    '@typescript-eslint',
  ],
  rules: {
    'no-param-reassign': ['error', {
      props: true,
      ignorePropertyModificationsFor: ['draft'],
    }],
    'no-use-before-define': 0,
    // 'platform-rules/no-src-import': [1, '@airslate/platform-*'],
    'react/button-has-type': 0,
    'react/destructuring-assignment': 0,
    'react/forbid-prop-types': 0,
    'react-hooks/rules-of-hooks': 'error',
    'react-hooks/exhaustive-deps': 'warn',
    'react/jsx-filename-extension': 0,
    'react/jsx-no-bind': [0],
    'react/jsx-one-expression-per-line': 0,
    'react/prefer-stateless-function': 0,
    'react/no-danger': 0,

    'import/extensions': 0,
    'import/no-extraneous-dependencies': 0,
    'import/no-unresolved': 0,
    'import/no-dynamic-require': 0,
    'import/prefer-default-export': 0,

    'jsx-a11y/click-events-have-key-events': 0,
    'jsx-a11y/no-noninteractive-element-interactions': 0,
    'jsx-a11y/label-has-associated-control': 0,
    'jsx-a11y/label-has-for': 0,
    'jsx-a11y/no-autofocus': 0,
    'jsx-a11y/href-no-hash': 0,
    'jsx-a11y/no-static-element-interactions': 0,

    'no-unused-expressions': 0,
    'no-useless-escape': 0,
    'no-underscore-dangle': 0,
    'no-mixed-operators': 0,
    'no-bitwise': 0,
    'no-prototype-builtins': 0,
    'no-path-concat': 0,
    'no-nested-ternary': 0,
    'no-sequences': 0,
    'no-unneeded-ternary': 0,
    'no-confusing-arrow': ['error', { allowParens: true }],
    'no-plusplus': 0,
    'no-constant-condition': ['error', { checkLoops: false }],

    'global-require': 0,
    'arrow-parens': 0,
    'operator-assignment': 0,
    indent: [2, 2, { SwitchCase: 1, flatTernaryExpressions: true }],

    'object-curly-newline': 0,
    'implicit-arrow-linebreak': 0,
    '@typescript-eslint/no-empty-function': 0,
    '@typescript-eslint/no-empty-interface': 0,
    '@typescript-eslint/ban-types': 0, // @TODO: uncomment
    '@typescript-eslint/ban-ts-comment': 0,
    '@typescript-eslint/no-inferrable-types': 'warn',
    '@typescript-eslint/no-explicit-any': 0,
    '@typescript-eslint/explicit-module-boundary-types': 0,
    '@typescript-eslint/no-non-null-assertion': 'error',
  },
};
